graph_components/GraphVariantAlignmentStats.cpp

```diff
@@ -0,0 +1,160 @@
+//
+// Expansion Hunter
+// Copyright 2016-2019 Illumina, Inc.
+// All rights reserved.
+//
+// Author: Egor Dolzhenko <edolzhenko@illumina.com>
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+//
+
+#include "graph_components/GraphVariantAlignmentStats.hh"
+
+#include <memory>
+
+#include <boost/algorithm/string/join.hpp>
+
+namespace ehunter
+{
+
+using graphtools::Graph;
+using graphtools::GraphAlignment;
+using graphtools::NodeId;
+using std::list;
+using std::string;
+using std::vector;
+
+static string encode(const vector<NodeId>& nodeIds)
+{
+    vector<string> encoding;
+    for (auto nodeId : nodeIds)
+    {
+        encoding.push_back(std::to_string(nodeId));
+    }
+
+    return boost::algorithm::join(encoding, ", ");
+}
+
+GraphVariantAlignmentStatsCalculator::GraphVariantAlignmentStatsCalculator(std::vector<graphtools::NodeId> variantNodes)
+    : variantNodes_(std::move(variantNodes))
+{
+    if (variantNodes_.empty())
+    {
+        throw std::logic_error("Cannot create a node bundle without nodes");
+    }
+
+    for (int index = 1; index != static_cast<int>(variantNodes_.size()); ++index)
+    {
+        if (variantNodes_[index] != variantNodes_[index - 1] + 1)
+        {
+            throw std::logic_error("Bundle " + encode(variantNodes_) + " must contain ordered and consecutive nodes");
+        }
+    }
+
+    firstVariantNode_ = variantNodes_.front();
+    lastVariantNode_ = variantNodes_.back();
+}
+
+void GraphVariantAlignmentStatsCalculator::inspect(const list<GraphAlignment>& alignments)
+{
+    for (const auto& alignment : alignments)
+    {
+        switch (classify(alignment))
+        {
+        case Flank::kLeft:
+            ++numReadsOverlappingLeftBreakpoint_;
+            return;
+        case Flank::kRight:
+            ++numReadsOverlappingRightBreakpoint_;
+            return;
+        case Flank::kBoth:
+            ++numReadsOverlappingLeftBreakpoint_;
+            ++numReadsOverlappingRightBreakpoint_;
+            return;
+        default:
+            break;
```

Comment:
So if the best alignment doesn't overlap either flank we want to keep going to the next one?

Reply:
That't right -- we are iterating over all top-scoring alignments until we find one that overlaps a flank. In practice, this should have a small/no effect on the coverage estimation.